### ISSUE TITLE
Early stopping callback 

### ISSUE THREAD

**Issue by adeboissiere** [PyTorchLightning/pytorch-lightning#2151]
## 🐛 Bug

Early stopping does not have the desired effect when creating a custom callback. Even when creating a custom callback with the default values, the training will stop before the early stopping before the conditions are met.

### To Reproduce

1. Create callback

```python
early_stop_callback = EarlyStopping(
        monitor='val_loss',
        min_delta=0.00,
        patience=3,
        verbose=False,
        mode='min',
        strict=True
    )
```

2. Create trainer

```python 
trainer = Trainer.from_argparse_args(Namespace(**dict(train_config)), early_stop_callback=early_stop_callback)
```

3. Train

```python
trainer.fit(model)
```

Here are the validation steps in the model:

```python
    def validation_step(self, batch, batch_idx):
        batch, y = batch
        y_hat = self(batch)

        loss = F.cross_entropy(y_hat, y.long())
        labels_hat = torch.argmax(y_hat, dim=1)
        n_correct_pred = torch.sum(y == labels_hat).item()

        return {'val_loss': loss, "n_correct_pred": n_correct_pred, "n_pred": len(y)}

    def validation_epoch_end(self, outputs):
        avg_loss = torch.stack([x['val_loss'] for x in outputs]).mean()
        val_acc = sum([x['n_correct_pred'] for x in outputs]) / sum(x['n_pred'] for x in outputs)
        tensorboard_logs = {'val_loss': avg_loss, 'val_acc': val_acc}

        return {'val_loss': avg_loss, 'log': tensorboard_logs}
```

### Expected behavior

In my case, training stops after 2 epochs, whether the validation loss increases or not. The callback behavior should be the same as the default. When I don't pass a custom callback, it works fine. I'm probably doing something wrong.

### Environment

 - PyTorch Version : 1.4.0+cu100
 - OS: Ubuntu 18.04
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Python version: 3.6.9
 - CUDA/cuDNN version: 10.0.130/7.6.4
 - GPU models and configuration: GeForce GTX 860M

Thanks'!

**Comment by Borda** [comment]
@jeremyjordan mind have a look ^^

**Comment by alekseynp** [comment]
What version of pytorch-lightning are you using?
I see a recent commit
https://github.com/PyTorchLightning/pytorch-lightning/commit/479ab49d03585d22f586e46f4334670a312f549b
Only available in master and `0.8.0rc1`
https://github.com/PyTorchLightning/pytorch-lightning/tree/0.8.0rc1

**Comment by alekseynp** [comment]
For the record, I arrived at this issue because in version 0.7.6 I observe early stopping not behaving properly.

**Comment by tpanum** [comment]
I have been puzzled about the issue of a CustomCallback never stopping in `max` mode (despite it clearly should've, considering the tensorboard logs). I am sitting on `0.7.6`. Thanks for pointing it out, @alekseynp. I'll see if an upgrade can fix it.

**Comment by DavidRuhe** [comment]
Hi. I'm not sure if I should create a separate issue, but there is a very confusing bug (still in the current master branch). The documentation states 

> By default early stopping will be enabled if ‘val_loss’ is found in validation_epoch_end()’s return dict. Otherwise training will proceed with early stopping disabled.

However, this is **not true** due to the following bug.
In callback_config.py we see the following code.

`    def configure_early_stopping(self, early_stop_callback):
        if early_stop_callback is True or None:
            self.early_stop_callback = EarlyStopping(
                monitor='val_loss',
                patience=3,
                strict=True,
                verbose=True,
                mode='min'
            )
            self.enable_early_stop = True
        elif not early_stop_callback:
            self.early_stop_callback = None
            self.enable_early_stop = False
        else:
            self.early_stop_callback = early_stop_callback
            self.enable_early_stop = True`

To trigger the behaviour as the documentation says it should be
`if early_stop_callback is True or early_stop_callback is None:`

David

**Comment by adeboissiere** [comment]
> What version of pytorch-lightning are you using?
> I see a recent commit
> [479ab49](https://github.com/PyTorchLightning/pytorch-lightning/commit/479ab49d03585d22f586e46f4334670a312f549b)
> Only available in master and `0.8.0rc1`
> https://github.com/PyTorchLightning/pytorch-lightning/tree/0.8.0rc1

Hi. I'm using version 0.7.6.

**Comment by jeremyjordan** [comment]
There are a number of issues with early stopping, I have a PR out to fix them. I have added [a new test](https://github.com/PyTorchLightning/pytorch-lightning/pull/1504/files#diff-a0606aadf6970149ee18c4edcbcfdc04R73) to cover your case @adeboissiere.

**Comment by brucemuller** [comment]
I'm also having issues. It seems to be only using the default values regardless and also the default values do not cause any early stopping to occur.

**Comment by williamFalcon** [comment]
Currently being worked on in #1504
